official/coin_build/man/man3/SbVec3s.3
.TH "SbVec3s" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SbVec3s \- The \fBSbVec3s\fP class is a 3 dimensional vector with short integer coordinates\&.
.PP
This vector class provides storage for a 3 dimensional vector as well as simple integer arithmetic operations\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/SbVec3s\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSbVec3s\fP (void)"
.br
.ti -1c
.RI "\fBSbVec3s\fP (const short v[3])"
.br
.ti -1c
.RI "\fBSbVec3s\fP (short x, short y, short z)"
.br
.ti -1c
.RI "\fBSbVec3s\fP (const \fBSbVec3us\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP (const \fBSbVec3b\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP (const \fBSbVec3i32\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP (const \fBSbVec3f\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP (const \fBSbVec3d\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBsetValue\fP (const short v[3])"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBsetValue\fP (short x, short y, short z)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBsetValue\fP (const \fBSbVec3us\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBsetValue\fP (const \fBSbVec3b\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBsetValue\fP (const \fBSbVec3i32\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBsetValue\fP (const \fBSbVec3f\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBsetValue\fP (const \fBSbVec3d\fP &v)"
.br
.ti -1c
.RI "const short * \fBgetValue\fP (void) const"
.br
.ti -1c
.RI "void \fBgetValue\fP (short &x, short &y, short &z) const"
.br
.ti -1c
.RI "short & \fBoperator[]\fP (\fBint\fP i)"
.br
.ti -1c
.RI "const short & \fBoperator[]\fP (\fBint\fP i) const"
.br
.ti -1c
.RI "int32_t \fBdot\fP (const \fBSbVec3s\fP &v) const"
.br
.ti -1c
.RI "void \fBnegate\fP (void)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBoperator*=\fP (\fBint\fP d)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBoperator*=\fP (double d)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBoperator/=\fP (\fBint\fP d)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBoperator/=\fP (double d)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBoperator+=\fP (const \fBSbVec3s\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP & \fBoperator\-=\fP (const \fBSbVec3s\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator\-\fP (void) const"
.br
.ti -1c
.RI "\fBSbString\fP \fBtoString\fP () const"
.br
.ti -1c
.RI "SbBool \fBfromString\fP (const \fBSbString\fP &str)"
.br
.ti -1c
.RI "void \fBprint\fP (FILE *fp) const"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "short \fBvec\fP [3]"
.br
.in -1c
.SS "Related Functions"
(Note that these are not member functions\&.) 
.in +1c
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator*\fP (const \fBSbVec3s\fP &v, \fBint\fP d)"
.br
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator*\fP (const \fBSbVec3s\fP &v, double d)"
.br
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator*\fP (\fBint\fP d, const \fBSbVec3s\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator*\fP (double d, const \fBSbVec3s\fP &v)"
.br
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator/\fP (const \fBSbVec3s\fP &v, \fBint\fP d)"
.br
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator/\fP (const \fBSbVec3s\fP &v, double d)"
.br
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator+\fP (const \fBSbVec3s\fP &v1, const \fBSbVec3s\fP &v2)"
.br
.ti -1c
.RI "\fBSbVec3s\fP \fBoperator\-\fP (const \fBSbVec3s\fP &v1, const \fBSbVec3s\fP &v2)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator==\fP (const \fBSbVec3s\fP &v1, const \fBSbVec3s\fP &v2)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!=\fP (const \fBSbVec3s\fP &v1, const \fBSbVec3s\fP &v2)"
.br
.in -1c
.SH "Detailed Description"
.PP 
The \fBSbVec3s\fP class is a 3 dimensional vector with short integer coordinates\&.
.PP
This vector class provides storage for a 3 dimensional vector as well as simple integer arithmetic operations\&. 

Be aware that this class is an extension for Coin, and it is not available in the original SGI Open Inventor v2\&.1 API\&.
.PP
\fBSee also:\fP
.RS 4
\fBSbVec2s\fP, \fBSbVec2f\fP, \fBSbVec2d\fP, \fBSbVec3f\fP, \fBSbVec3d\fP, \fBSbVec4f\fP, \fBSbVec4d\fP\&. 
.RE
.PP
\fBSince:\fP
.RS 4
Coin 2\&.0 
.PP
TGS Inventor ?\&.? 
.RE
.PP

.SH "Constructor & Destructor Documentation"
.PP 
.SS "SbVec3s::SbVec3s (void)\fC [inline]\fP"
The default constructor does nothing\&. The vector coordinates will be uninitialized until you do a \fBsetValue()\fP call\&. 
.SS "SbVec3s::SbVec3s (const short v[3])\fC [inline]\fP"
Constructs an \fBSbVec3s\fP instance with initial values from \fIv\fP\&. 
.SS "SbVec3s::SbVec3s (short x, short y, short z)\fC [inline]\fP"
Constructs an \fBSbVec3s\fP instance with the initial vector endpoint set to \fI<x\fP,y,z>\&. 
.SS "SbVec3s::SbVec3s (const \fBSbVec3us\fP & v)\fC [inline]\fP, \fC [explicit]\fP"

.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "SbVec3s::SbVec3s (const \fBSbVec3b\fP & v)\fC [inline]\fP, \fC [explicit]\fP"

.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "SbVec3s::SbVec3s (const \fBSbVec3i32\fP & v)\fC [inline]\fP, \fC [explicit]\fP"

.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "SbVec3s::SbVec3s (const \fBSbVec3f\fP & v)\fC [inline]\fP, \fC [explicit]\fP"

.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "SbVec3s::SbVec3s (const \fBSbVec3d\fP & v)\fC [inline]\fP, \fC [explicit]\fP"

.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "\fBSbVec3s\fP & SbVec3s::setValue (const short v[3])\fC [inline]\fP"
Set new coordinates for the vector from \fIv\fP\&. Returns reference to self\&.
.PP
\fBSee also:\fP
.RS 4
\fBgetValue()\fP\&. 
.RE
.PP

.SS "\fBSbVec3s\fP & SbVec3s::setValue (const \fBSbVec3us\fP & v)"
\fBSbVec3s\fP & SbVec3s::setValue(short x, short y, short z)
.PP
Set new coordinates for the vector\&. Returns reference to self\&.
.PP
\fBSee also:\fP
.RS 4
\fBgetValue()\fP\&.
.RE
.PP
Sets the value from an \fBSbVec3us\fP instance\&.
.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "\fBSbVec3s\fP & SbVec3s::setValue (const \fBSbVec3b\fP & v)"
Sets the value from an \fBSbVec3b\fP instance\&.
.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "\fBSbVec3s\fP & SbVec3s::setValue (const \fBSbVec3i32\fP & v)"
Sets the value from an \fBSbVec3i32\fP instance\&.
.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "\fBSbVec3s\fP & SbVec3s::setValue (const \fBSbVec3f\fP & v)"
Sets the value from an \fBSbVec3f\fP instance\&.
.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "\fBSbVec3s\fP & SbVec3s::setValue (const \fBSbVec3d\fP & v)"
Sets the value from an \fBSbVec3d\fP instance\&.
.PP
\fBSince:\fP
.RS 4
Coin 2\&.5 
.RE
.PP

.SS "const short * SbVec3s::getValue (void) const\fC [inline]\fP"
Returns a pointer to an array of three shorts containing the x, y and z coordinates of the vector\&.
.PP
\fBSee also:\fP
.RS 4
\fBsetValue()\fP\&. 
.RE
.PP

.SS "void SbVec3s::getValue (short & x, short & y, short & z) const\fC [inline]\fP"
Returns the x, y and z coordinates of the vector\&.
.PP
\fBSee also:\fP
.RS 4
\fBsetValue()\fP\&. 
.RE
.PP

.SS "short & SbVec3s::operator[] (\fBint\fP i)\fC [inline]\fP"
Index operator\&. Returns modifiable x, y or z coordinate of vector\&.
.PP
\fBSee also:\fP
.RS 4
\fBgetValue()\fP and \fBsetValue()\fP\&. 
.RE
.PP

.SS "short SbVec3s::operator[] (\fBint\fP i) const\fC [inline]\fP"
Index operator\&. Returns x, y or z coordinate of vector\&.
.PP
\fBSee also:\fP
.RS 4
\fBgetValue()\fP and \fBsetValue()\fP\&. 
.RE
.PP

.SS "int32_t SbVec3s::dot (const \fBSbVec3s\fP & v) const\fC [inline]\fP"
Calculates and returns the result of taking the dot product of this vector and \fIv\fP\&. 
.SS "void SbVec3s::negate (void)\fC [inline]\fP"
Negate the vector (i\&.e\&. point it in the opposite direction)\&. 
.SS "\fBSbVec3s\fP & SbVec3s::operator*= (\fBint\fP d)\fC [inline]\fP"
Multiply components of vector with value \fId\fP\&. Returns reference to self\&. 
.SS "\fBSbVec3s\fP & SbVec3s::operator*= (double d)"
Multiply components of vector with value \fId\fP\&. Returns reference to self\&. 
.SS "\fBSbVec3s\fP & SbVec3s::operator/= (\fBint\fP d)\fC [inline]\fP"
Divides components of vector with value \fId\fP\&. Returns reference to self\&. 
.SS "\fBSbVec3s\fP & SbVec3s::operator/= (double d)\fC [inline]\fP"
Divides components of vector with value \fId\fP\&. Returns reference to self\&. 
.SS "\fBSbVec3s\fP & SbVec3s::operator+= (const \fBSbVec3s\fP & v)\fC [inline]\fP"
Adds this vector and vector \fIv\fP\&. Returns reference to self\&. 
.SS "\fBSbVec3s\fP & SbVec3s::operator\-= (const \fBSbVec3s\fP & v)\fC [inline]\fP"
Subtracts vector \fIv\fP from this vector\&. Returns reference to self\&. 
.SS "\fBSbVec3s\fP SbVec3s::operator\- (void) const\fC [inline]\fP"
Non-destructive negation operator\&. Returns a new \fBSbVec3s\fP instance which points in the opposite direction of this vector\&.
.PP
\fBSee also:\fP
.RS 4
\fBnegate()\fP\&. 
.RE
.PP

.SS "\fBSbString\fP SbVec3s::toString () const"
Return a string representation of this object 
.SS "SbBool SbVec3s::fromString (const \fBSbString\fP & str)"
Convert from a string representation, return wether this is a valid conversion 
.SS "void SbVec3s::print (FILE * fp) const"
Dump the state of this object to the \fIfile\fP stream\&. Only works in debug version of library, method does nothing in an optimized compile\&. 
.SH "Friends And Related Function Documentation"
.PP 
.SS "\fBSbVec3s\fP operator* (const \fBSbVec3s\fP & v, \fBint\fP d)\fC [related]\fP"
Returns an \fBSbVec3s\fP instance which is the components of vector \fIv\fP multiplied with \fId\fP\&. 
.SS "\fBSbVec3s\fP operator* (const \fBSbVec3s\fP & v, double d)\fC [related]\fP"
Returns an \fBSbVec3s\fP instance which is the components of vector \fIv\fP multiplied with \fId\fP\&. 
.SS "\fBSbVec3s\fP operator* (\fBint\fP d, const \fBSbVec3s\fP & v)\fC [related]\fP"
Returns an \fBSbVec3s\fP instance which is the components of vector \fIv\fP multiplied with \fId\fP\&. 
.SS "\fBSbVec3s\fP operator* (double d, const \fBSbVec3s\fP & v)\fC [related]\fP"
Returns an \fBSbVec3s\fP instance which is the components of vector \fIv\fP multiplied with \fId\fP\&. 
.SS "\fBSbVec3s\fP operator/ (const \fBSbVec3s\fP & v, \fBint\fP d)\fC [related]\fP"
Returns an \fBSbVec3s\fP instance which is the components of vector \fIv\fP divided on the scalar factor \fId\fP\&. 
.SS "\fBSbVec3s\fP operator/ (const \fBSbVec3s\fP & v, double d)\fC [related]\fP"
Returns an \fBSbVec3s\fP instance which is the components of vector \fIv\fP divided on \fId\fP\&. 
.SS "\fBSbVec3s\fP operator+ (const \fBSbVec3s\fP & v1, const \fBSbVec3s\fP & v2)\fC [related]\fP"
Returns an \fBSbVec3s\fP instance which is the sum of vectors \fIv1\fP and \fIv2\fP\&. 
.SS "\fBSbVec3s\fP operator\- (const \fBSbVec3s\fP & v1, const \fBSbVec3s\fP & v2)\fC [related]\fP"
Returns an \fBSbVec3s\fP instance which is vector \fIv2\fP subtracted from vector \fIv1\fP\&. 
.SS "\fBint\fP operator== (const \fBSbVec3s\fP & v1, const \fBSbVec3s\fP & v2)\fC [related]\fP"
Returns \fI1\fP if \fIv1\fP and \fIv2\fP are \fIexactly\fP equal, \fI0\fP otherwise\&. 
.SS "\fBint\fP operator!= (const \fBSbVec3s\fP & v1, const \fBSbVec3s\fP & v2)\fC [related]\fP"
Returns \fI1\fP if \fIv1\fP and \fIv2\fP are not equal, \fI0\fP if they are equal\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/base/SbVec3s.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

#include <Inventor/SbVec3s.h>

#include <limits>
#include <cassert>

#include <Inventor/SbVec3us.h>
#include <Inventor/SbVec3b.h>
#include <Inventor/SbVec3i32.h>
#include <Inventor/SbVec3f.h>
#include <Inventor/SbVec3d.h>
#include <Inventor/fields/SoSFVec3s.h>
#if COIN_DEBUG
#include <Inventor/errors/SoDebugError.h>
#endif // COIN_DEBUG

#include "coinString.h"

/*!
  \class SbVec3s SbLinear.h Inventor/SbLinear.h
  \brief The SbVec3s class is a 3 dimensional vector with short integer
  coordinates.
  \ingroup base

  This vector class provides storage for a 3 dimensional vector
  as well as simple integer arithmetic operations.

  \COIN_CLASS_EXTENSION

  \sa SbVec2s, SbVec2f, SbVec2d, SbVec3f, SbVec3d, SbVec4f, SbVec4d.
  \since Coin 2.0
  \since TGS Inventor ?.?
*/

/*!
  \fn SbVec3s::SbVec3s(void)

  The default constructor does nothing. The vector coordinates will be
  uninitialized until you do a setValue() call.
*/

/*!
  \fn SbVec3s::SbVec3s(const short v[3])

  Constructs an SbVec3s instance with initial values from \a v.
*/

/*!
  \fn SbVec3s::SbVec3s(short x, short y, short z)

  Constructs an SbVec3s instance with the initial vector endpoint set to
  \a <x,y,z>.
*/

/*!
  \fn SbVec3s::SbVec3s(const SbVec3us & v)

  \since Coin 2.5
*/

/*!
  \fn SbVec3s::SbVec3s(const SbVec3b & v)

  \since Coin 2.5
*/

/*!
  \fn SbVec3s::SbVec3s(const SbVec3i32 & v)

  \since Coin 2.5
*/

/*!
  \fn SbVec3s::SbVec3s(const SbVec3f & v)

  \since Coin 2.5
*/

/*!
  \fn SbVec3s::SbVec3s(const SbVec3d & v)

  \since Coin 2.5
*/

/*!
  \fn int32_t SbVec3s::dot(const SbVec3s & v) const

  Calculates and returns the result of taking the dot product of this
  vector and \a v.
*/

/*!
  \fn const short * SbVec3s::getValue(void) const

  Returns a pointer to an array of three shorts containing the x, y
  and z coordinates of the vector.

  \sa setValue().
*/

/*!
  \fn void SbVec3s::getValue(short & x, short & y, short & z) const

  Returns the x, y and z coordinates of the vector.

  \sa setValue().
*/

/*!
  \fn void SbVec3s::negate(void)

  Negate the vector (i.e. point it in the opposite direction).
*/

/*!
  \fn SbVec3s & SbVec3s::setValue(const short v[3])

  Set new coordinates for the vector from \a v. Returns reference to
  self.

  \sa getValue().
*/

/*!
  SbVec3s & SbVec3s::setValue(short x, short y, short z)

  Set new coordinates for the vector. Returns reference to self.

  \sa getValue().
*/

/*!
  Sets the value from an SbVec3us instance.

  \since Coin 2.5
*/

SbVec3s &
SbVec3s::setValue(const SbVec3us & v)
{
  vec[0] = static_cast<short>(v[0]);
  vec[1] = static_cast<short>(v[1]);
  vec[2] = static_cast<short>(v[2]);
  return *this;
}

/*!
  Sets the value from an SbVec3b instance.

  \since Coin 2.5
*/

SbVec3s &
SbVec3s::setValue(const SbVec3b & v)
{
  vec[0] = static_cast<short>(v[0]);
  vec[1] = static_cast<short>(v[1]);
  vec[2] = static_cast<short>(v[2]);
  return *this;
}

/*!
  Sets the value from an SbVec3i32 instance.

  \since Coin 2.5
*/

SbVec3s &
SbVec3s::setValue(const SbVec3i32 & v)
{
#if COIN_DEBUG
  if (v[0] > std::numeric_limits<short>::max() || v[0] < -std::numeric_limits<short>::max() ||
      v[1] > std::numeric_limits<short>::max() || v[1] < -std::numeric_limits<short>::max() ||
      v[2] > std::numeric_limits<short>::max() || v[2] < -std::numeric_limits<short>::max()) {
    SoDebugError::post("SbVec3s::setValue", "SbVec3i32 argument out of range for SbVec3s");
  }
#endif // COIN_DEBUG
  vec[0] = static_cast<short>(v[0]);
  vec[1] = static_cast<short>(v[1]);
  vec[2] = static_cast<short>(v[2]);
  return *this;
}

/*!
  Sets the value from an SbVec3f instance.

  \since Coin 2.5
*/

SbVec3s &
SbVec3s::setValue(const SbVec3f & v)
{
#if COIN_DEBUG
  if (v[0] > std::numeric_limits<short>::max() || v[0] < -std::numeric_limits<short>::max() ||
      v[1] > std::numeric_limits<short>::max() || v[1] < -std::numeric_limits<short>::max() ||
      v[2] > std::numeric_limits<short>::max() || v[2] < -std::numeric_limits<short>::max()) {
    SoDebugError::post("SbVec3s::setValue", "SbVec3f argument out of range for SbVec3s");
  }
#endif // COIN_DEBUG
  vec[0] = static_cast<short>(v[0]);
  vec[1] = static_cast<short>(v[1]);
  vec[2] = static_cast<short>(v[2]);
  return *this;
}

/*!
  Sets the value from an SbVec3d instance.

  \since Coin 2.5
*/

SbVec3s &
SbVec3s::setValue(const SbVec3d & v)
{
#if COIN_DEBUG
  if (v[0] > std::numeric_limits<short>::max() || v[0] < -std::numeric_limits<short>::max() ||
      v[1] > std::numeric_limits<short>::max() || v[1] < -std::numeric_limits<short>::max() ||
      v[2] > std::numeric_limits<short>::max() || v[2] < -std::numeric_limits<short>::max()) {
    SoDebugError::post("SbVec3s::setValue", "SbVec3d argument out of range for SbVec3s");
  }
#endif // COIN_DEBUG
  vec[0] = static_cast<short>(v[0]);
  vec[1] = static_cast<short>(v[1]);
  vec[2] = static_cast<short>(v[2]);
  return *this;
}

/*!
  \fn short & SbVec3s::operator[](const int i)

  Index operator. Returns modifiable x, y or z coordinate of vector.

  \sa getValue() and setValue().
*/

/*!
  \fn short SbVec3s::operator[](const int i) const
  Index operator. Returns x, y or z coordinate of vector.

  \sa getValue() and setValue().
*/

/*!
  \fn SbVec3s & SbVec3s::operator *= (int d)
  Multiply components of vector with value \a d. Returns reference to self.
*/

/*!
  Multiply components of vector with value \a d. Returns reference to self.
*/

SbVec3s &
SbVec3s::operator *=(double d)
{
  vec[0] = static_cast<short>(vec[0] * d);
  vec[1] = static_cast<short>(vec[1] * d);
  vec[2] = static_cast<short>(vec[2] * d);
  return *this;
}

/*!
  \fn SbVec3s & SbVec3s::operator /= (int d)

  Divides components of vector with value \a d. Returns reference to self.
*/

/*!
  \fn SbVec3s & SbVec3s::operator /= (double d)

  Divides components of vector with value \a d. Returns reference to self.
*/

/*!
  \fn SbVec3s & SbVec3s::operator += (const SbVec3s & v)

  Adds this vector and vector \a v. Returns reference to self.
*/

/*!
  \fn SbVec3s & SbVec3s::operator -= (const SbVec3s & v)
  Subtracts vector \a v from this vector. Returns reference to self.
*/

/*!
  \fn SbVec3s SbVec3s::operator-(void) const

  Non-destructive negation operator. Returns a new SbVec3s instance
  which points in the opposite direction of this vector.

  \sa negate().
*/

/*!
  \fn SbVec3s operator * (const SbVec3s & v, int d)
  \relates SbVec3s

  Returns an SbVec3s instance which is the components of vector \a v
  multiplied with \a d.
*/

/*!
  \fn SbVec3s operator * (const SbVec3s & v, double d)
  \relates SbVec3s

  Returns an SbVec3s instance which is the components of vector \a v
  multiplied with \a d.
*/

/*!
  \fn SbVec3s operator * (int d, const SbVec3s & v)
  \relates SbVec3s

  Returns an SbVec3s instance which is the components of vector \a v
  multiplied with \a d.
*/

/*!
  \fn SbVec3s operator * (double d, const SbVec3s & v)
  \relates SbVec3s

  Returns an SbVec3s instance which is the components of vector \a v
  multiplied with \a d.
*/

/*!
  \fn SbVec3s operator / (const SbVec3s & v, int d)
  \relates SbVec3s

  Returns an SbVec3s instance which is the components of vector \a v
  divided on the scalar factor \a d.
*/

/*!
  \fn SbVec3s operator / (const SbVec3s & v, double d)
  \relates SbVec3s

  Returns an SbVec3s instance which is the components of vector \a v
  divided on \a d.
*/

/*!
  \fn SbVec3s operator + (const SbVec3s & v1, const SbVec3s & v2)
  \relates SbVec3s

  Returns an SbVec3s instance which is the sum of vectors \a v1 and \a v2.
*/

/*!
  \fn SbVec3s operator - (const SbVec3s & v1, const SbVec3s & v2)
  \relates SbVec3s

  Returns an SbVec3s instance which is vector \a v2 subtracted from
  vector \a v1.
*/

/*!
  \fn int operator == (const SbVec3s & v1, const SbVec3s & v2)
  \relates SbVec3s

  Returns \a 1 if \a v1 and \a v2 are \e exactly equal, \a 0 otherwise.
*/

/*!
  \fn int operator != (const SbVec3s & v1, const SbVec3s & v2)
  \relates SbVec3s

  Returns \a 1 if \a v1 and \a v2 are not equal, \a 0 if they are equal.
*/

/*!
  Return a string representation of this object
*/
SbString
SbVec3s::toString() const
{
  return CoinInternal::ToString(*this);
}

/*!
  Convert from a string representation, return wether this is a valid conversion
*/
bool
SbVec3s::fromString(const SbString & str)
{
  bool conversionOk;
  *this = CoinInternal::FromString<SbVec3s>(str,&conversionOk);
  return conversionOk;
}

/*!
  Dump the state of this object to the \a file stream. Only works in
  debug version of library, method does nothing in an optimized
  compile.
*/
void
SbVec3s::print(FILE * fp) const
{
#if COIN_DEBUG
  SbString str = "<";
  str+=this->toString();
  str += ">";
  fputs(str.getString(),fp);
#endif // COIN_DEBUG
}

#ifdef COIN_TEST_SUITE
#include <Inventor/SbTypeInfo.h>

typedef SbVec3s ToTest;
BOOST_AUTO_TEST_CASE(toString) {
  ToTest val(1,2,3);
  SbString str("1 2 3");
  BOOST_CHECK_MESSAGE(str == val.toString(),
                      std::string("Mismatch between ") +  val.toString().getString() + " and control string " + str.getString());
}

BOOST_AUTO_TEST_CASE(fromString) {
  ToTest foo;
  SbString test = "1 -2 3";
  ToTest trueVal(1,-2,3);
  foo.fromString(test);
  BOOST_CHECK_MESSAGE(trueVal == foo,
                      std::string("Mismatch between ") +  foo.toString().getString() + " and control " + trueVal.toString().getString());
}

BOOST_AUTO_TEST_CASE(fromInvalidString) {
  ToTest foo;
  SbString test = "a,2,3";
  bool conversionOk = foo.fromString(test);
  BOOST_CHECK_MESSAGE(conversionOk == false,
                      std::string("Able to convert from ") + test.getString() + " which is not a valid " + SbTypeInfo<ToTest>::getTypeName() + " representation");
}

#endif //COIN_TEST_SUITE
